hw/ip/pwm/data/pwm.hjson

```diff
@@ -0,0 +1,133 @@
+          name: "CNTR_EN",
+          desc: '''Assert this bit to enable the PWM counter.
+                   Writing 0 disables and resets the PWM counter.'''
+          resval: "0x0"
+        },
+        { bits: "15:0",
+          name: "CLK_DIV",
+          desc: '''Sets the base period for the rest of the IP.
+                   The output drive frequency for each channel will be the input clock frequency, divided by 65536*(CLK_DIV+1).
+                   The factor 65536 accounts for the fact that the internal phase counters operate with 16-bit resolution.'''
+          resval: "0x00000090"
+        }
+      ]
+    },
+    { multireg: { name: "PWM_EN",
+        desc: "Enable PWM operation for each channel",
+        count: "NOutputs",
+        swaccess: "rw",
+        cname: "pwm_en",
+        compact: "1",
+        fields: [
+          { bits: "0",
+            name: "EN",
+            desc: '''Write 1 to this bit to enable PWM pulses on the corresponding channel.''',
+            resval: "0"
+          }
+        ]
+      }
+    },
+    { multireg: { name: "PWM_PARAM",
+        desc: "Controls the duty cycle and phase of each channel.",
+        count: "NOutputs"
+        swaccess: "rw",
+        cname: "pwm_param",
+        fields: [
+          { bits: "31:16",
+            name: "PHASE_DELAY",
+            desc: "Phase delay of the PWM rising edge, in units of 1/65536 of a cycle",
+            resval: "0x0000"
+          },
+          { bits: "15:0",
+            name: "DUTY_CYCLE",
+            desc: '''PWM output duty cycle.  A value of 0 corresponds to the minimal duty cycle of 1/65536.
+                     The total duty cycle is (DUTY_CYCLE+1)/65536'''
+            resval: "0x7fff"
+          }
+        ]
+      }
+    },
+    { multireg: { name: "BLINK_EN",
+        desc: "Enable Hardware controlled blink feature for each channel",
+        count: "NOutputs"
+        swaccess: "rw",
+        cname: "blink_en",
+        compact: "1",
+        fields: [
+          { bits: "0",
+            name: "EN",
+            desc: '''Write 1 to this bit to enable blinking for the corresponding channel.
+                     Clearing this bit to zero disables blinking, meaning the corresponding
+                     channel will receive continuous pulses.  Clearing this bit also resets
+                     the internal blink counters.'''
+            resval: "0"
+          }
+        ]
+      }
+    },
+    { multireg: { name: "BLINK_PARAM",
+        desc: "Hardware controlled blink parameters.",
+        count: "NOutputs"
+        swaccess: "rw",
+        cname: "blink_param",
+        fields: [
+          { bits: "31:24",
+            name: "PREDIV",
+            desc: '''Channel-specific blink frequency pre-divider.  The blink frequency for each channel is given by
+                     the common PWM drive frequency divided by (RATE+1)*(PREDIV+1).'''
+            resval: "0x00"
+          }
+          { bits: "23:16",
```

Comment:
i read the description in `_index.md`, but i'm not quite following the need for both a rate and a duty cycle.
Do you mind elaborating? 

I think a diagram would be helpful for me.

Reply:
Duty cycle is the *fraction* of time the device is on.  (Not sure why people use the word "cycle" in this context, but it's pretty standard).     Rate and Pre-div together define the pulse rate.    In a perfect world one would just have one divider, but I was thinking that if the input clock were 100s of MHz it would be silly to require a 24- or 32-bit field for duty cycle. 

However the blink settings have a variable phase resolution (which tops out at 2<sup>8</sup>) I could simplify this all by:
- Giving the PWM variable phase-resolution would allow the input clock to be arbitrarily slow, and would make the blink control and PWM control registers look the same.
- Making all the clock dividers 24- or 32-bit (and getting rid of pre-dividers) would conceptually simplify this all.   
    - In this case the blink and PWM delay and duty-cycle registers would have the same resolution as the corresponding clock divider.

I can certainly add a diagram.

Reply:
ah having them the same programmers model does sound appealing.  And it sounds like the number of bits all come out to be about the same?

Reply:
Indeed a diagram to show the path from clock to PWM with all the different divisors will help.

Reply:
Diagram added.

Reply:
Looks good. Maybe it worth to add a clock input as well.